raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java

```diff
@@ -0,0 +1,489 @@
+    @Test
+    void testAddressWithEmptyEndpoint() {
+        Endpoints endpoints = Endpoints.empty();
+
+        Optional<InetSocketAddress> address = endpoints.address(ListenerName.normalised("nonExistentListener"));
+
+        assertEquals(Optional.empty(), address);
+    }
+
+    @Test
+    void testVotersRecordEndpointsWithEndpoint() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        endpointMap.put(
+                ListenerName.normalised("listener"),
+                InetSocketAddress.createUnresolved("localhost", 9092)
+        );
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+        Iterator<VotersRecord.Endpoint> iterator = endpoints.votersRecordEndpoints();
+
+        VotersRecord.Endpoint endpoint = iterator.next();
+
+        assertEquals(ListenerName.normalised("listener").value(), endpoint.name());
+        assertEquals("localhost", endpoint.host());
+        assertEquals(9092, endpoint.port());
+    }
+
+    @Test
+    void testVotersRecordEndpointsWithEmptyEndpoint() {
+        Endpoints endpoints = Endpoints.empty();
+
+        Iterator<VotersRecord.Endpoint> iterator = endpoints.votersRecordEndpoints();
+
+        assertFalse(iterator.hasNext());
+    }
+
+    @Test
+    void testSizeWithOneEndpoint() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        endpointMap.put(
+                ListenerName.normalised("listener"),
+                InetSocketAddress.createUnresolved("localhost", 9092)
+        );
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        int size = endpoints.size();
+
+        assertEquals(1, size);
+    }
+
+    @Test
+    void testSizeWithEmptyEndpoint() {
+        Endpoints endpoints = Endpoints.empty();
+
+        int size = endpoints.size();
+
+        assertEquals(0, size);
+    }
+
+    @Test
+    void testIsEmptyWithEndpoint() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        endpointMap.put(
+                ListenerName.normalised("listener"),
+                InetSocketAddress.createUnresolved("localhost", 9092)
+        );
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        assertFalse(endpoints.isEmpty());
+    }
+
+    @Test
+    void testIsEmptyWithEmptyEndpoint() {
+        Endpoints endpoints = Endpoints.empty();
+
+        assertEquals(0, endpoints.size());
+    }
+
+    @Test
+    void testEqualsAndHashCodeWithSameEndpoint() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        endpointMap.put(
+                ListenerName.normalised("listener"),
+                InetSocketAddress.createUnresolved("localhost", 9092)
+        );
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        Endpoints sameEndpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        assertEquals(endpoints, sameEndpoints);
```

Comment:
Am not sure what the value of this test? can you explain it a bit as I recall we rarely test `hashCode` method.

Reply:
Don't we need to make sure the overridden `equals()` method performs correctly? Happy to remove it is not needed.

Reply:
Seems reasonable to test `equals`